File: src/main/java/org/jcodec/containers/mp4/boxes/SampleSizesBox.java
package org.jcodec.containers.mp4.boxes;

import java.nio.ByteBuffer;

import org.jcodec.containers.mp4.boxes.Box.AtomField;

/**
 * This class is part of JCodec ( www.jcodec.org )
 * This software is distributed under FreeBSD License
 * 
 * @author The JCodec project
 *
 */
public class SampleSizesBox extends FullBox {
    public SampleSizesBox(Header atom) {
        super(atom);
    }

    private int defaultSize;
    private int count;
    private int[] sizes;
    
    public static String fourcc() {
        return "stsz";
    }

    public static SampleSizesBox createSampleSizesBox(int defaultSize, int count) {
        SampleSizesBox stsz = new SampleSizesBox(new Header(fourcc()));
        stsz.defaultSize = defaultSize;
        stsz.count = count;
        return stsz;
    }

    public static SampleSizesBox createSampleSizesBox2(int[] sizes) {
        SampleSizesBox stsz = new SampleSizesBox(new Header(fourcc()));
        stsz.sizes = sizes;
        stsz.count = sizes.length;
        return stsz;
    }

    public void parse(ByteBuffer input) {
        super.parse(input);
        defaultSize = input.getInt();
        count = input.getInt();

        if (defaultSize == 0) {
            sizes = new int[count];
            for (int i = 0; i < count; i++) {
                sizes[i] = input.getInt();
            }
        }
    }

    @AtomField(idx=0)
    public int getDefaultSize() {
        return defaultSize;
    }

    @AtomField(idx=2)
    public int[] getSizes() {
        return sizes;
    }
    
    @AtomField(idx=1)
    public int getCount() {
        return count;
    }
    
    public void setCount(int count) {
        this.count = count;
    }

    @Override
    public void doWrite(ByteBuffer out) {
        super.doWrite(out);
        out.putInt((int) defaultSize);

        if (defaultSize == 0) {
            out.putInt(count);
            for (int i = 0; i < sizes.length; i++) {
                long size = sizes[i];
                out.putInt((int) size);
            }
        } else {
            out.putInt((int)count);
        }
    }
    
    @Override
    public int estimateSize() {
        return (defaultSize == 0 ? sizes.length * 4 : 0) + 20;
    }

    public void setSizes(int[] sizes) {
        this.sizes = sizes;
        this.count = sizes.length;
    }
}

File: src/main/java/org/jcodec/containers/mp4/boxes/SampleSizesBox.kt
package org.jcodec.containers.mp4.boxes

import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org )
 * This software is distributed under FreeBSD License
 *
 * @author The JCodec project
 */
class SampleSizesBox(atom: Header) : FullBox(atom) {
    var defaultSize = 0
        private set
    private var count = 0
    private var sizes: IntArray = IntArray(0)
    override fun parse(input: ByteBuffer) {
        super.parse(input)
        defaultSize = input.int
        count = input.int
        if (defaultSize == 0) {
            sizes = IntArray(count)
            for (i in 0 until count) {
                sizes[i] = input.int
            }
        }
    }

    fun getSizes(): IntArray {
        return sizes
    }

    fun getCount(): Int {
        return count
    }

    fun setCount(count: Int) {
        this.count = count
    }

    public override fun doWrite(out: ByteBuffer) {
        super.doWrite(out)
        out.putInt(defaultSize)
        if (defaultSize == 0) {
            out.putInt(count)
            for (i in sizes.indices) {
                val size = sizes[i].toLong()
                out.putInt(size.toInt())
            }
        } else {
            out.putInt(count)
        }
    }

    override fun estimateSize(): Int {
        return (if (defaultSize == 0) sizes.size * 4 else 0) + 20
    }

    fun setSizes(sizes: IntArray) {
        this.sizes = sizes
        count = sizes.size
    }

    companion object {
        @JvmStatic
        fun fourcc(): String {
            return "stsz"
        }

        @JvmStatic
        fun createSampleSizesBox(defaultSize: Int, count: Int): SampleSizesBox {
            val stsz = SampleSizesBox(Header(fourcc()))
            stsz.defaultSize = defaultSize
            stsz.count = count
            return stsz
        }

        @JvmStatic
        fun createSampleSizesBox2(sizes: IntArray): SampleSizesBox {
            val stsz = SampleSizesBox(Header(fourcc()))
            stsz.sizes = sizes
            stsz.count = sizes.size
            return stsz
        }
    }
}